File: api/src/main/java/mb/pie/api/Share.java
package mb.pie.api;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.function.Supplier;

/**
 * Share for concurrently executing tasks.
 */
public interface Share {
    TaskData share(TaskKey key, Supplier<TaskData> execFunc, @Nullable Supplier<TaskData> visitedFunc);
}

File: runtime/src/main/kotlin/mb/pie/runtime/Share.kt
package mb.pie.runtime

/**
 * Share for concurrently executing tasks.
 */
interface Share {
  fun reuseOrCreate(task: UTask, cacheFunc: (UTask) -> UTaskData?, execFunc: (UTask) -> UTaskData): UTaskData
  fun reuseOrCreate(task: UTask, execFunc: (UTask) -> UTaskData): UTaskData
}
